sched/scheduler/task_scheduler.go

```diff
@@ -9,27 +9,49 @@ type taskAssignments struct {
 	task *taskState
 }
 
-// Returns a list of taskAssigments of task to available node.  Not all
-// tasks are guaranteed to be scheduled.  Does best effort scheduling
-func getTaskAssignments(cs *clusterState, tasks []*taskState) []taskAssignments {
+// Returns a list of taskAssigments of task to available node.
+// Also returns a modified copy of clusterState.nodeGroups for the caller to apply (so this remains a pure fn).
+// Does best effort scheduling which tries to assign tasks to nodes already primed for similar tasks.
+// Not all tasks are guaranteed to be scheduled.
+func getTaskAssignments(cs *clusterState, tasks []*taskState) ([]taskAssignments, map[string]*nodeGroup) {
+	var assignments []taskAssignments
+	nodeGroups := map[string]*nodeGroup{}
+	snapshotIds := []string{}
 
-	// for now just assign any task to any available node
-	var tas []taskAssignments
-	taskIndex := 0
+	// Create a copy of nodeGroups to modify based on new scheduling.
+	for snapId, groups := range cs.nodeGroups {
+		nodeGroups[snapId] = newNodeGroup()
+		for nodeId, node := range groups.idle {
+			nodeGroups[snapId].idle[nodeId] = node
+		}
+		for nodeId, node := range groups.busy {
+			nodeGroups[snapId].busy[nodeId] = node
+		}
+		snapshotIds = append(snapshotIds, snapId)
+	}
 
-	for _, nodeState := range cs.nodes {
-		if nodeState.runningTask == noTask {
-			if taskIndex < len(tasks) {
-				tas = append(tas, taskAssignments{
-					node: nodeState.node,
-					task: tasks[taskIndex],
-				})
-				taskIndex++
-			} else {
-				// we've scheduled all tasks
-				break
+	// Loop over all snapshotIDs looking for an idle node. Prefer, in order:
+	// - Hot node for the given snapshotID (one whose last task shared the same snapshotID).
+	// - New untouched node (or node whose last task used an empty snapshotID)
+	// - Any node from the idle pools of nodes associated with other snapshotIDs.
+Loop:
+	for _, task := range tasks {
```

Comment:
I'm probably going to punt on this unless the following applies:

I suggest we set our criteria for code quality such that we're not spending time on what could be ascribed to preference. We should accept code where:
1) Intent is clear
2) Behavior isn't suspicious wrt unintended side effects

We've spent a considerable # of dev hours in the past doing ultimately unnecessary things like runner refactor and going forward I'd like to avoid allocating any time that way, even for something as small as this section of code.

What does everyone else think?